rfcs/rfc-0001-lookup-by-username.md

```diff
@@ -0,0 +1,51 @@
+- Feature Name: (API route to look up user by username)
+- Start Date: (15/03/2023)
+- RFC PR: https://github.com/revoltchat/rfcs/issues/1
+- Tracking Issue: https://github.com/revoltchat/revolt/issues/0000
+- Status: draft
+
+# Summary
+
+Add an API route to look up users by username instead of ID
+
+# Motivation
+
+Currently, bots need to keep a cache of every user if they want to allow commands to target users by username (e.g. /ban @insert).
+I am proposing an API route that allows clients to look up users by username.
+
+# Guide-level explanation
+
+Send a `GET` request to `/users/by-name/insert` (no idea how to name the route) to get an array of users matching that username.
+
+- Lookup should be case insensitive
+- The response should be an array, in case multiple users managed to get the same username assigned somehow and to future proof for when discriminators will be added
+
+# Reference-level explanation
+
+I don't know
+
+# Drawbacks
+
+- Might require another index in MongoDB
+- Since the backend needs to make sure the returned users are mutual, there might be more processing overhead involved, but the logic for this can probably be copied from the /users/:id route
```

Comment:
so this means you can only look up users that you already have some form of connection to? via friends, a group dm, or a server you're both in, etc